Add support for Azure LB Standard and Public IP Standard

justaugustus opened the pull request:
Ref: #579

justaugustus commented:
@tombuildsstuff would you mind taking a peep at this?

tombuildsstuff commented on azurerm/resource_arm_loadbalancer.go:
```diff
@@ -114,6 +119,7 @@ func resourceArmLoadBalancerCreate(d *schema.ResourceData, meta interface{}) err
 	name := d.Get("name").(string)
 	location := d.Get("location").(string)
 	resGroup := d.Get("resource_group_name").(string)
+	sku := convertSkuStringToLoadBalancerSku(d.Get("sku").(string))
```
rather than parsing this out I think we can make this `network.LoadBalancerSkuName(d.Get("sku").(string))`?

tombuildsstuff commented on azurerm/resource_arm_loadbalancer.go:
```diff
@@ -100,6 +100,11 @@ func resourceArmLoadBalancer() *schema.Resource {
 
+			"sku": {
+				Type:     schema.TypeString,
+				Required: true,
```
from what I can tell existing Load Balancers are of the Basic SKU; so we can instead default this to `Basic` and make this Optional; like so:

```
Default: string(network.LoadBalancerSkuNameBasic),
Optional: true,
```

^ this will mean that existing users don't need to specify the SKU (or update the configuration) for any existing Load Balancers. We can also extend this to add Validation and allow mixed-case strings (e.g. `Basic` and `basic`) via:

```
ValidateFunc: validation.StringInSlice([]string{
  string(network.LoadBalancerSkuNameBasic),
  string(network.LoadBalancerSkuNameStandard),
}, true),
DiffSupressFunc: ignoreCaseDiffSuppressFunc,
```

justaugustus commented:
@tombuildsstuff Thanks so much for the quick CR!

I've addressed your notes and have a few follow-up questions...
1. The tests for basic and standard have the same resource names for `azurerm_lb` (`test`). Will that cause issues during testing?
2. To support public LB Standard resources, we'll also need to make a similar addition of the `sku` parameter to the `azurerm_public_ip` resource. Should I include that change with this PR or send another one?

tombuildsstuff commented:
@justaugustus

> I've addressed your notes and have a few follow-up questions...

> 1. The tests for basic and standard have the same resource names for azurerm_lb (test). Will that cause issues during testing?

Each of the acceptance tests suffixes the resource names (in this case the LB Name) with a random integer/string to make the name unique across tests - so there shouldn't be any issues here.

> 2. To support public LB Standard resources, we'll also need to make a similar addition of the sku parameter to the azurerm_public_ip resource. Should I include that change with this PR or send another one?

Ideally it should be split out into another PR (which we can merge before this one) - but given they're interdependent I don't see any harm combining them into one if that's easier :)

Thanks!

justaugustus commented:
I'll send another PR for the Public IP Standard SKU stuff.
Thanks again for the feedback, @tombuildsstuff!

JunyiYi commented:
@justaugustus , thanks for your contribution.
@tombuildsstuff , CoreOS is trying to understand when this feature will be supported in Terraform. Maybe you can give the date after this pull request is merged.

Thanks.

justaugustus commented:
Hey @tombuildsstuff! 
Changed my mind and decided to bundle the two SKUs in this PR (seemed simpler that way).

I added acceptance tests, but I was thinking through a few additional cases and a little unsure about how to approach them:

### LB
#### Test SKU:
- basic -> PASS
- standard -> FAIL
- OTHER -> FAIL

#### Test frontend IP config:
- private -> PASS
- public:
    - PIP SKU:
        - basic:
            - LB SKU:
                - basic -> PASS
                - standard -> FAIL
        - standard:
            - LB SKU:
                - basic -> FAIL
                - standard -> PASS

### PIP
#### Test SKU:
- basic -> PASS
- standard:
    - allocation:
        - static -> PASS
        - dynamic -> FAIL
        - OTHER -> FAIL


Could you take another sweep through this?

paultyng commented on azurerm/resource_arm_loadbalancer_test.go:
```diff
@@ -209,6 +228,27 @@ resource "azurerm_lb" "test" {
+resource "azurerm_lb" "test" {
+    name = "arm-test-loadbalancer-%d"
+    location = "${azurerm_resource_group.test.location}"
+		resource_group_name = "${azurerm_resource_group.test.name}"
```
got some mixed tabs and spaces here

justaugustus replied:
@paultyng I'm noticing this across multiple files in the provider... are tabs or spaces preferred?

paultyng commented on azurerm/resource_arm_public_ip.go:
```diff
@@ -84,6 +84,15 @@ func resourceArmPublicIp() *schema.Resource {
+				Optional:         true,
+				Default:          string(network.PublicIPAddressSkuNameBasic),
+				ForceNew:         true,
+				ValidateFunc:     validatePublicIpSku,
```
This could just be `StringInSlice` couldn't it?

justaugustus replied:
@paultyng I was mimicking the other validation functions in that file and figured it could be reused in the acceptance tests, similar to:
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_public_ip_test.go#L41
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_public_ip_test.go#L73

Is there a preferred approach?

tombuildsstuff replied:
It depends what's being tested - but in general we're using the Constants/Enums in the SDK via the `StringInSlice` (at least for Strings) for newer resources; given these constants exist - could we switch this for a `StringInSlice`?

paultyng commented on azurerm/resource_arm_public_ip_test.go:
```diff
@@ -318,6 +338,23 @@ resource "azurerm_public_ip" "test" {
+    name = "acctestpublicip-%d"
+    location = "${azurerm_resource_group.test.location}"
+    resource_group_name = "${azurerm_resource_group.test.name}"
+		public_ip_address_allocation = "static"
```
mixed tabs and spaces

justaugustus replied:
@paultyng I'm noticing this across multiple files in the provider... are tabs or spaces preferred?

paultyng commented on website/docs/r/public_ip.html.markdown:
```diff
@@ -42,7 +42,11 @@ The following arguments are supported:
-* `public_ip_address_allocation` - (Required) Defines whether the IP address is stable or dynamic. Options are Static or Dynamic.
+* `sku` - (Optional) The SKU of the Public IP. Accepted values are `Basic` and `Standard`. Defaults to `Basic`.
+
+~> **Note** Public IP Standard SKUs require `public_ip_address_allocation` to be set to `static`.
```
Should this be in a validation?

tombuildsstuff replied:
@justaugustus apologies for the delayed response - we've been trying to get the SDKv12 migration finished up asap (which turned out to be a considerably larger job than we initially anticipated).

At the current time it's only possible to access the value of that specific field in the validation method on the Schema - as such I believe we'd need to do this in the `resourceArmPublicIpCreate` method?

tombuildsstuff replied:
I've added a commit to explain that this feature is still in Preview and needs to be opted-in

justaugustus commented:
@tombuildsstuff / @paultyng --
thanks for the feedback! I've addressed both of your reviews, so this is ripe for another sweep when you have a chance. :)

tombuildsstuff commented:
@justaugustus hey - just to let you know that now the SDK upgrade work is completed that I'm hoping to take a look into this later today (apologies for the delay here - we've been trying to minimise merge conflicts during the SDK migration) :)

tombuildsstuff commented on azurerm/resource_arm_public_ip.go:
```diff
@@ -185,6 +210,16 @@ func resourceArmPublicIpRead(d *schema.ResourceData, meta interface{}) error {
+	}
+
+	if resp.PublicIPAddressPropertiesFormat.DNSSettings != nil && resp.PublicIPAddressPropertiesFormat.DNSSettings.Fqdn != nil && *resp.PublicIPAddressPropertiesFormat.DNSSettings.Fqdn != "" {
+		d.Set("fqdn", resp.PublicIPAddressPropertiesFormat.DNSSettings.Fqdn)
```
we can remove this line, since it's set below: https://github.com/terraform-providers/terraform-provider-azurerm/pull/665/files#diff-ed59d26a8bebf9bf3b188b200b02dd27L193

tombuildsstuff commented on website/docs/r/loadbalancer.html.markdown:
```diff
@@ -45,6 +45,7 @@ The following arguments are supported:
 * `resource_group_name` - (Required) The name of the resource group in which to create the LoadBalancer.
 * `location` - (Required) Specifies the supported Azure location where the resource exists.
 * `frontend_ip_configuration` - (Optional) A frontend ip configuration block as documented below.
+* `sku` - (Optional) The SKU of the Azure Load Balancer. Accepted values are `Basic` and `Standard`. Defaults to `Basic`.
```
I've added a commit to explain that this feature is still in Preview and needs to be opted-in

tombuildsstuff commented:
Tests pass for both LB's and Public IP's after registering for the Preview:

<img width="466" alt="screen shot 2018-01-26 at 16 02 13" src="https://user-images.githubusercontent.com/666005/35448530-6f1ba694-02b2-11e8-8379-383e32e0584c.png">

<img width="456" alt="screen shot 2018-01-26 at 16 02 17" src="https://user-images.githubusercontent.com/666005/35448534-71e0e330-02b2-11e8-991b-ea528db7d4b0.png">

justaugustus commented:
:tada:
